### ISSUE TITLE
Detect variable scoping & declaration issues

### ISSUE THREAD

**Issue by jakearchibald** [tabatkins/bikeshed#1255]
Could bikeshed detect when a 'variable' is used before it's declared, declared multiple times, or if it's used out-of-scope, or perhaps even alert when it has a scope wider than it needs?

I guess this would need an opt-in like `<div algorithm strict-vars>`.

Then a way to distinguish between variable usage and declaration, such as `<var init>foo</var>`, and a shorthand like `|=foo=|`.

Variables would be scoped to their parent `<ol>` or `<div algorithm>` as per [infra](https://infra.spec.whatwg.org/#variables).

**Comment by tabatkins** [comment]
Hmm, I'm curious about this "block scoped" notion here. Just linking to the Wikipedia definition doesn't really help, given that HTML doesn't have an obvious notion of "blocks". I'll bug anne about this.

**Comment by jakearchibald** [comment]
Yeah, that's fair. File it with infra. I've always assumed this meant parent "ol", but it'd be good to clarify what happens with "dl"s and "ul"s.

**Comment by jakearchibald** [comment]
https://github.com/whatwg/infra/issues/204

**Comment by jakearchibald** [comment]
A "v1" of this feature could assume that the first usage of a variable within an algorithm is its declaration, and associate it with the closest ol/algorithm. Then, bikeshed can check additional uses of variables of that name, and if it's out of scope, warn.

This wouldn't need `strict-vars` or a syntax for declarations.
